File: 981.TimeBasedKey-ValueStore.py
# Initail Solution

class TimeMap:

    def __init__(self):
        self.records = [] # (time stamp, {key:value})

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.records.append((timestamp, {key:value}))
        
    def get(self, key: str, timestamp: int) -> str:
        l, r = 0, len(self.records) - 1
        while l < r:
            mid = (l + r) // 2
            if self.records[mid][0] == timestamp:
                return self.records[mid][1][key]
            elif self.records[mid][0] > timestamp:
                r = mid - 1
            else:
                l = mid + 1
        for i in range(l, -1, -1):
            if self.records[i][0] <= timestamp and key in self.records[i][1].keys():
                return self.records[i][1][key]
        
        return "" 
 


class TimeMap:

    def __init__(self):
        self.records = {} # (value, timestamp)

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.records.keys():
            self.records[key] = []
        self.records[key].append([value, timestamp])

    def get(self, key: str, timestamp: int) -> str:
        res = ""
        if key not in self.records.keys():
            return res

        curRecord = self.records[key]
        l, r = 0 , len(curRecord) - 1
        while l <= r:
            mid = (l + r) // 2
            if curRecord[mid][1] <= timestamp:
                res = curRecord[mid][0]
                l = mid + 1
            else:
                r = mid - 1
        return res

File: 981.TimeBasedKey-ValueStore.java
class Pair{
    int timestamp;
    String value;
    public Pair(String value, int timestamp){
        this.value = value;
        this.timestamp = timestamp;
    }
}
class TimeMap {
    HashMap<String, List<Pair>> map;
    public TimeMap() {
        map = new HashMap<>();
    }
    
    public void set(String key, String value, int timestamp) {
        if(!map.containsKey(key)){
            map.put(key, new ArrayList<>());
        }
        map.get(key).add(new Pair(value, timestamp));
    }
    
    public String get(String key, int timestamp) {
        if(!map.containsKey(key)){
            return "";
        }else{
            return binarySearch(map.get(key), timestamp);
        }
    }

    public String binarySearch(List<Pair> arr, int timestamp){
        int low = 0, high = arr.size() - 1;
        String res = "";

        while (low <= high){
            int mid = (low + high) / 2;
            if(arr.get(mid).timestamp <= timestamp){
                res = arr.get(mid).value;
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }
        return res;
    }
}
